README.md
# Multipart Related
Fast parsing `multipart/related` responses in the browser or node. Operates on raw bytes, in a streaming manner. Zero dependencies.

This is a research project and not meant to be production ready (yet).

Multipart construction is planned for the future, too, therefore the generic name.


## API
The package exports a `MultipartRelated` constructor:

### `MultipartRelated(contentType)`
`contentType` must be a `multipart/related` Content-Type string including the boundary, as usually sent by the servers.
Eg `'multipart/related; boundary="865f3b787e6623728e6aa49fec037303"` or `'multipart/related; boundary=865f3b787e6623728e6aa49fec037303`.

This class provides a `read(chunk)` function, which inteded to be used in conjunction with [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream), as you can see in the example above. It returns an array containing the parsed parts, which are already present.

A part is an object containing
* `boundary`: if the content is grouped together, aka _related_, this includes the individual boundary. If the part is a sole one, `boundary` is `null`. Can be used to group together parts.
* `headers`: the headers of each individual parts, usually eg `Content-Type`, `Content-Disposition`, `Content-Length` or `Content-Encoding` but not limited to those.
* `data`: an `Uint8Array` containing the raw part data.


## Example
```js
import MultipartRelated from 'multipart-related'

// fetch the multipart response
const response = await fetch(url)

// get Content-Type header
const contentType = response.headers.get('Content-Type')

// initialize the multipart parser
const multipart = new MultipartRelated(contentType)

// get the response body stream
const reader = response.body.getReader()

// tie together the process stream
const process = ({ value, done }) => {
  const parts = multipart.read(value)
  // do something with the parts as they arrive
  console.log(parts)

  if (!done) {
    return reader.read().then(process)
  }
}
// ... and kick-off processing
reader.read().then(process)
```


## Tests
Run the tests with `npm test` or pick individual files via eg `node test/multipart-related-parser-test.js`.


(c) 2022 Johannes J. Schmidt


CHANGELOG.md
# Changelog

## 2.0.0
* **Breaking change:** parts `related` property has been renamed to `boundary`
* **fix:** under some conditions when split into multiple chunks not all parts have been parsed

## 1.0.0
Initial version


test/index-test.js
import tap from 'tap'
import fs from 'fs'
import path from 'path'

import MultipartRelated from '../src/index.js'

tap.test('doc', t => {
  t.plan(8)

  const contentType = 'multipart/related; boundary="865f3b787e6623728e6aa49fec037303"'
  const unpacker = new MultipartRelated(contentType)

  const filename = path.join(process.cwd(), 'test/fixtures/doc.multipart')

  return fs.readFile(filename, (error, data) => {
    // read in two chunks
    const parts = unpacker.read(data.slice(0, 1000))
      .concat(unpacker.read(data.slice(1000)))
    
    t.equal(parts.length, 4, 'correct number of parts parsed')

    const [doc, three, two, one] = parts
    
    t.same(doc.headers, {
      'Content-Type': 'application/json'
    }, 'correct headers parsed for doc')

    t.same(three.headers, {
      'Content-Disposition': 'attachment; filename="three.txt"',
      'Content-Type': 'text/plain',
      'Content-Length': 69,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for three')
    t.equal(three.data.length, 69, 'correct length of data returned')

    t.same(two.headers, {
      'Content-Disposition': 'attachment; filename="two.txt"',
      'Content-Type': 'text/plain',
      'Content-Length': 59,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for two')
    t.equal(two.data.length, 59, 'correct length of data returned')

    t.same(one.headers, {
      'Content-Disposition': 'attachment; filename="one.txt"',
      'Content-Type': 'text/plain',
      'Content-Length': 50,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for one')
    t.equal(one.data.length, 50, 'correct length of data returned')
  })
})

tap.test('bulk get', t => {
  t.plan(12)

  const contentType = 'multipart/related; boundary="--440d45a70251a521a20f0930e9a5420e"'
  const unpacker = new MultipartRelated(contentType)

  const filename = path.join(process.cwd(), 'test/fixtures/_bulk_get.multipart')

  return fs.readFile(filename, (error, data) => {
    const parts = unpacker.read(data)
    
    t.equal(parts.length, 7)
    
    const [docOne, three, two, one, docTwo, tf, cool] = parts

    t.same(docOne.headers, {
      'Content-Type': 'application/json'
    }, 'correct headers parsed for doc')

    t.same(three.headers, {
      'Content-Disposition': 'attachment; filename="three.txt"',
      'Content-Type': 'text/plain',
      'Content-Length': 69,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for three')
    t.equal(three.data.length, 69, 'correct length of data returned')

    t.same(two.headers, {
      'Content-Disposition': 'attachment; filename="two.txt"',
      'Content-Type': 'text/plain',
      'Content-Length': 59,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for two')
    t.equal(two.data.length, 59, 'correct length of data returned')

    t.same(one.headers, {
      'Content-Disposition': 'attachment; filename="one.txt"',
      'Content-Type': 'text/plain',
      'Content-Length': 50,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for one')
    t.equal(one.data.length, 50, 'correct length of data returned')

    t.same(tf.headers, {
      'Content-Disposition': 'attachment; filename="tf.gif"',
      'Content-Type': 'image/gif',
      'Content-Length': 1450
    }, 'correct headers parsed for tf')
    t.equal(tf.data.length, 1450, 'correct length of data returned')

    t.same(cool.headers, {
      'Content-Disposition': 'attachment; filename="cool"',
      'Content-Type': 'text/plain',
      'Content-Length': 24,
      'Content-Encoding': 'gzip'
    }, 'correct headers parsed for cool')
    t.equal(cool.data.length, 24, 'correct length of data returned')
  })
})


test/first-boundary-position.js
import tap from 'tap'

import firstBoundaryPosition from '../src/first-boundary-position.js'

tap.test('smoke', t => {
  const data = [1,45,45,2,3]
  const boundary = [2,3]
  const index = firstBoundaryPosition(data, boundary)
  t.equal(index, 1)
  t.end()
})

tap.test('offset bigger than data', t => {
  const data = [1,2,3]
  const boundary = [2,3]
  const index = firstBoundaryPosition(data, boundary, 3)
  t.equal(index, -1)
  t.end()
})

tap.test('offset bigger than data plus boundary', t => {
  const data = [1,2,3]
  const boundary = [3,4]
  const index = firstBoundaryPosition(data, boundary, 1)
  t.equal(index, -1)
  t.end()
})

tap.test('half match', t => {
  const data = [1,45,45,2,3]
  const boundary = [2,4]
  const index = firstBoundaryPosition(data, boundary)
  t.equal(index, -1)
  t.end()
})


test/multipart-related-parser-test.js
import tap from 'tap'

import MultipartRelatedParser from '../src/multipart-related-parser.js'

const encoder = new TextEncoder()

tap.test('single part', t => {
  const parser = new MultipartRelatedParser('multipart/related; boundary=boundary')

  const text = `--boundary\r
Content-Type: text/plain\r
\r
text one\r
--boundary--`
  const data = encoder.encode(text)

  const part = parser.parsePart(data)
  t.same(part, {
    boundary: null,
    headers: {
      'Content-Type': 'text/plain'
    },
    data: encoder.encode('text one'),
    rest: new Uint8Array(0)
  })

  t.end()
})


tap.test('two parts', t => {
  const parser = new MultipartRelatedParser('multipart/related; boundary=boundary')

  const text = `--boundary\r
Content-Type: text/plain\r
\r
text one\r
--boundary\r
Content-Type: text/plain\r
\r
text number two\r
--boundary--`
  const data = encoder.encode(text)

  const one = parser.parsePart(data)
  t.equal(one.boundary, null, 'related not set')
  t.same(one.headers, {
    'Content-Type': 'text/plain'
  }, 'correct headers parsed')
  t.same(one.data, encoder.encode('text one'), 'correct part data extracted')
  t.same(one.rest, encoder.encode(`--boundary\r
Content-Type: text/plain\r
\r
text number two\r
--boundary--`), 'correct data rest present')

  const two = parser.parsePart(one.rest)
  t.equal(two.boundary, null, 'related not set')
  t.same(two.headers, {
    'Content-Type': 'text/plain'
  }, 'correct headers parsed')
  t.same(two.data, encoder.encode('text number two'), 'correct part data extracted')
  t.same(two.rest, new Uint8Array(0), 'empty data rest present')

  t.end()
})

tap.test('using content length', t => {
  const parser = new MultipartRelatedParser('multipart/related; boundary=boundary')

  const text = `--boundary\r
Content-Type: text/plain\r
Content-Length: 12\r
\r
text one and a half\r
--boundary--`
  const data = encoder.encode(text)

  const part = parser.parsePart(data)
  t.equal(part.boundary, null, 'related not set')
  t.same(part.headers, {
    'Content-Type': 'text/plain',
    'Content-Length': 12
  }, 'correct headers parsed')
  t.same(part.data, encoder.encode('text one and a half'), 'correct part data extracted, regardless of content length')
  t.same(part.rest, new Uint8Array(0), 'empty data rest present')

  t.end()
})

tap.test('two parts using wrong content length', t => {
  const parser = new MultipartRelatedParser('multipart/related; boundary=boundary')

  const text = `--boundary\r
Content-Type: text/plain\r
Content-Length: 32\r
\r
text one\r
--boundary\r
Content-Type: text/plain\r
\r
text number two\r
--boundary--`
  const data = encoder.encode(text)

  const one = parser.parsePart(data)
  t.equal(one.boundary, null, 'related not set')
  t.same(one.headers, {
    'Content-Type': 'text/plain',
    'Content-Length': 32
  }, 'correct headers parsed')
  t.same(one.data, encoder.encode(`text one\r
--boundary\r
Content-Type: text/plain\r
\r
text number two`), 'correct part data extracted due to content length')
  t.same(one.rest, new Uint8Array(0), 'empty data rest present')

  const two = parser.parsePart(one.rest)
  t.notOk(two, 'There is no part two')

  t.end()
})

tap.test('test related content', t => {
  const parser = new MultipartRelatedParser('multipart/related; boundary=boundary')

  const text = `--boundary\r
Content-Type: multipart/related; boundary=child-boundary\r
\r
--child-boundary\r
Content-Type: text/plain\r
\r
text one\r
--child-boundary--\r
--boundary\r
Content-Type: text/plain\r
\r
text number two\r
--boundary--`
  const data = encoder.encode(text)

  const one = parser.parsePart(data)
  t.equal(one.boundary, 'child-boundary', 'related is set')
  t.same(one.headers, {
    'Content-Type': 'text/plain'
  }, 'correct headers parsed')
  t.same(one.data, encoder.encode('text one'), 'correct part data extracted')
  t.same(one.rest, encoder.encode(`--boundary\r
Content-Type: text/plain\r
\r
text number two\r
--boundary--`), 'correct data rest present')

  const two = parser.parsePart(one.rest)
  t.equal(two.boundary, null, 'related is not set')
  t.same(two.headers, {
    'Content-Type': 'text/plain'
  }, 'correct headers parsed')
  t.same(two.data, encoder.encode('text number two'), 'correct part data extracted')
  t.same(two.rest, new Uint8Array(0), 'empty data rest present')

  t.end()
})


src/index.js
// Unpacking a multipart response

import MultipartRelatedParser from './multipart-related-parser.js'

// Parse a binary multipart/related stream
export default class MultipartRelated {
  constructor (contentType) {
    this.parser = new MultipartRelatedParser(contentType)
    this.data = new Uint8Array(0)
  }

  read (chunk) {
    if (chunk) {
      // add current chunk to data
      const newData = new Uint8Array(this.data.length + chunk.length)
      newData.set(this.data, 0)
      newData.set(chunk, this.data.length)
      this.data = newData
    }

    const parts = []

    // parse buffer
    let part
    do {
      part = this.parser.parsePart(this.data)
      if (part) {
        const { boundary, headers, data, rest } = part
        this.data = rest
        parts.push({
          boundary,
          headers,
          data
        })
      }
    } while (part)

    // return parsed parts
    return parts
  }
}


wasm_gzip.js
let wasm,cachedTextDecoder=new TextDecoder("utf-8",{ignoreBOM:!0,fatal:!0});cachedTextDecoder.decode();let cachegetUint8Memory0=null;function getUint8Memory0(){return null!==cachegetUint8Memory0&&cachegetUint8Memory0.buffer===wasm.memory.buffer||(cachegetUint8Memory0=new Uint8Array(wasm.memory.buffer)),cachegetUint8Memory0}function getStringFromWasm0(e,t){return cachedTextDecoder.decode(getUint8Memory0().subarray(e,e+t))}const heap=new Array(32).fill(void 0);heap.push(void 0,null,!0,!1);let heap_next=heap.length;function addHeapObject(e){heap_next===heap.length&&heap.push(heap.length+1);const t=heap_next;return heap_next=heap[t],heap[t]=e,t}function getObject(e){return heap[e]}function dropObject(e){e<36||(heap[e]=heap_next,heap_next=e)}function takeObject(e){const t=getObject(e);return dropObject(e),t}let WASM_VECTOR_LEN=0,cachedTextEncoder=new TextEncoder("utf-8");const encodeString="function"==typeof cachedTextEncoder.encodeInto?function(e,t){return cachedTextEncoder.encodeInto(e,t)}:function(e,t){const n=cachedTextEncoder.encode(e);return t.set(n),{read:e.length,written:n.length}};function passStringToWasm0(e,t,n){if(void 0===n){const n=cachedTextEncoder.encode(e),r=t(n.length);return getUint8Memory0().subarray(r,r+n.length).set(n),WASM_VECTOR_LEN=n.length,r}let r=e.length,a=t(r);const o=getUint8Memory0();let s=0;for(;s<r;s++){const t=e.charCodeAt(s);if(t>127)break;o[a+s]=t}if(s!==r){0!==s&&(e=e.slice(s)),a=n(a,r,r=s+3*e.length);const t=getUint8Memory0().subarray(a+s,a+r);s+=encodeString(e,t).written}return WASM_VECTOR_LEN=s,a}function isLikeNone(e){return null==e}let cachegetInt32Memory0=null;function getInt32Memory0(){return null!==cachegetInt32Memory0&&cachegetInt32Memory0.buffer===wasm.memory.buffer||(cachegetInt32Memory0=new Int32Array(wasm.memory.buffer)),cachegetInt32Memory0}function passArray8ToWasm0(e,t){const n=t(1*e.length);return getUint8Memory0().set(e,n/1),WASM_VECTOR_LEN=e.length,n}function getArrayU8FromWasm0(e,t){return getUint8Memory0().subarray(e/1,e/1+t)}export function compressGzip(e){try{const o=wasm.__wbindgen_add_to_stack_pointer(-16);var t=passArray8ToWasm0(e,wasm.__wbindgen_malloc),n=WASM_VECTOR_LEN;wasm.compressGzip(o,t,n);var r=getInt32Memory0()[o/4+0],a=getInt32Memory0()[o/4+1];let s;return 0!==r&&(s=getArrayU8FromWasm0(r,a).slice(),wasm.__wbindgen_free(r,1*a)),s}finally{wasm.__wbindgen_add_to_stack_pointer(16)}}export function decompressGzip(e){try{const o=wasm.__wbindgen_add_to_stack_pointer(-16);var t=passArray8ToWasm0(e,wasm.__wbindgen_malloc),n=WASM_VECTOR_LEN;wasm.decompressGzip(o,t,n);var r=getInt32Memory0()[o/4+0],a=getInt32Memory0()[o/4+1];let s;return 0!==r&&(s=getArrayU8FromWasm0(r,a).slice(),wasm.__wbindgen_free(r,1*a)),s}finally{wasm.__wbindgen_add_to_stack_pointer(16)}}export function compressStringGzip(e){try{const o=wasm.__wbindgen_add_to_stack_pointer(-16);var t=passStringToWasm0(e,wasm.__wbindgen_malloc,wasm.__wbindgen_realloc),n=WASM_VECTOR_LEN;wasm.compressStringGzip(o,t,n);var r=getInt32Memory0()[o/4+0],a=getInt32Memory0()[o/4+1];let s;return 0!==r&&(s=getArrayU8FromWasm0(r,a).slice(),wasm.__wbindgen_free(r,1*a)),s}finally{wasm.__wbindgen_add_to_stack_pointer(16)}}export function decompressStringGzip(e){try{const o=wasm.__wbindgen_add_to_stack_pointer(-16);var t=passArray8ToWasm0(e,wasm.__wbindgen_malloc),n=WASM_VECTOR_LEN;wasm.decompressStringGzip(o,t,n);var r=getInt32Memory0()[o/4+0],a=getInt32Memory0()[o/4+1];let s;return 0!==r&&(s=getStringFromWasm0(r,a).slice(),wasm.__wbindgen_free(r,1*a)),s}finally{wasm.__wbindgen_add_to_stack_pointer(16)}}let stack_pointer=32;function addBorrowedObject(e){if(1==stack_pointer)throw new Error("out of js stack");return heap[--stack_pointer]=e,stack_pointer}export function compressJsonGzip(e){try{const a=wasm.__wbindgen_add_to_stack_pointer(-16);wasm.compressJsonGzip(a,addBorrowedObject(e));var t=getInt32Memory0()[a/4+0],n=getInt32Memory0()[a/4+1],r=getArrayU8FromWasm0(t,n).slice();return wasm.__wbindgen_free(t,1*n),r}finally{wasm.__wbindgen_add_to_stack_pointer(16),heap[stack_pointer++]=void 0}}export function decompressJsonGzip(e){var t=passArray8ToWasm0(e,wasm.__wbindgen_malloc),n=WASM_VECTOR_LEN;return takeObject(wasm.decompressJsonGzip(t,n))}function handleError(e){return function(){try{return e.apply(this,arguments)}catch(e){wasm.__wbindgen_exn_store(addHeapObject(e))}}}async function load(e,t){if("function"==typeof Response&&e instanceof Response){if("function"==typeof WebAssembly.instantiateStreaming)try{return await WebAssembly.instantiateStreaming(e,t)}catch(t){if("application/wasm"==e.headers.get("Content-Type"))throw t;console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n",t)}const n=await e.arrayBuffer();return await WebAssembly.instantiate(n,t)}{const n=await WebAssembly.instantiate(e,t);return n instanceof WebAssembly.Instance?{instance:n,module:e}:n}}async function init(e){void 0===e&&(e=import.meta.url.replace(/\.js$/,"_bg.wasm"));const t={wbg:{}};t.wbg.__wbindgen_string_new=function(e,t){return addHeapObject(getStringFromWasm0(e,t))},t.wbg.__wbg_parse_58b7cdbfa2b3e55a=handleError((function(e,t){return addHeapObject(JSON.parse(getStringFromWasm0(e,t)))})),t.wbg.__wbindgen_object_drop_ref=function(e){takeObject(e)},t.wbg.__wbg_stringify_e5f075a4462d77f0=handleError((function(e){return addHeapObject(JSON.stringify(getObject(e)))})),t.wbg.__wbindgen_string_get=function(e,t){const n=getObject(t);var r="string"==typeof n?n:void 0,a=isLikeNone(r)?0:passStringToWasm0(r,wasm.__wbindgen_malloc,wasm.__wbindgen_realloc),o=WASM_VECTOR_LEN;getInt32Memory0()[e/4+1]=o,getInt32Memory0()[e/4+0]=a},t.wbg.__wbindgen_throw=function(e,t){throw new Error(getStringFromWasm0(e,t))},t.wbg.__wbindgen_rethrow=function(e){throw takeObject(e)},("string"==typeof e||"function"==typeof Request&&e instanceof Request||"function"==typeof URL&&e instanceof URL)&&(e=fetch(e));const{instance:n,module:r}=await load(await e,t);return wasm=n.exports,init.__wbindgen_wasm_module=r,wasm}export default init;

src/multipart-related-parser.js
import firstBoundaryPosition from './first-boundary-position.js'
import firstHeaderSeparatorPosition from './first-header-separator-position.js'
import startsWithBoundaryEnd from './starts-with-boundary-end.js'

export default class MultipartRelatedParser {
  constructor (contentType) {
    this.encoder = new TextEncoder()
    this.decoder = new TextDecoder()

    const boundary = this.parseContentType(contentType)
    this.id = boundary.id
    this.boundaries = [
      boundary
    ]
  }

  parseContentType (contentType) {
    const [_, type, id] = contentType.match(/^([^;]+);\s*boundary="?([^="]+)"?/) || []
    if (type !== 'multipart/related') return

    const boundary = this.encoder.encode(id)
    
    return {
      id,
      boundary
    }
  }

  parsePart (data) {
    if (this.boundaries.length === 0) return null

    const { id, boundary } = this.boundaries[this.boundaries.length - 1]

    // each part starts with the boundary marker
    const startPosition = firstBoundaryPosition(data, boundary)
    if (startPosition === -1) return null

    // find the end of the header, which is the start of the content
    const contentPosition = firstHeaderSeparatorPosition(data, startPosition)
    if (contentPosition === -1) return null

    // parse the header
    const headerData = data.slice(boundary.length + 4, contentPosition)
    const header = this.decoder.decode(headerData)
    const headers = header.split('\r\n').reduce((memo, line) => {
      const [name, value] = line.split(/:\s*/)
      memo[name] = name === 'Content-Length'
        ? parseInt(value, 10)
        : value
      return memo
    }, {})

    // content type is mandatory
    const { 'Content-Type': contentType } = headers
    if (!contentType) return null

    // check for content length, otherwise search for next boundary
    const { 'Content-Length': contentLength } = headers
    const contentEndPosition = contentLength
      ? firstBoundaryPosition(data, boundary, contentLength + contentPosition + 6)
      : firstBoundaryPosition(data, boundary, contentPosition + 4)
    if (contentEndPosition === -1) return null

    // part not completely present yet
    if (data.length < contentEndPosition) return null

    // check whether this starts a related part
    const childBoundary = this.parseContentType(contentType)
    if (childBoundary) {
      this.boundaries.push(childBoundary)
      return this.parsePart(data.slice(contentPosition + 4))
    }

    // check for boundary end marker
    const isBoundaryEnd = startsWithBoundaryEnd(data, boundary, contentEndPosition)
    const endPosition = isBoundaryEnd
      ? contentEndPosition + boundary.length + 6
      : contentEndPosition

    if (isBoundaryEnd) this.boundaries.pop()
  
    // extract the part data
    const partData = data.slice(contentPosition + 4, contentEndPosition - 2)
    const rest = data.slice(endPosition)

    return {
      boundary: this.id === id ? null : id,
      headers,
      data: partData,
      rest
    }
  }
}


multipart-unpacker.js
// Unpacking a multipart response

// use a WASM implementation for decompression: https://github.com/ColinTimBarndt/wasm-gzip
// until DecompressionStream is implemented widely
import init, { decompressGzip } from "./wasm_gzip.js"

// Separator between part header and its content: "\r\n\r\n"
const HEADER_SEPARATOR = new Uint8Array([13, 10, 13, 10])

// get first position of a binary match
const firstMatch = (matcher, data, offset = 0) => {
  if (offset > data.length) {
    return -1
  }

  for (var i = offset; i < data.length; i++) {
    if (data[i] === matcher[0]) {
      for (var j = i + 1, k = 1; j < i + matcher.length; j++, k++) {
        var fullMatch = true
        if (data[j] !== matcher[k]) {
          fullMatch = false
          break
        }
      }
      if (fullMatch) return j - matcher.length
    }
  }

  return -1
}

// Parse a binary multipart/related stream, eg:
// ```
// --7c5ea2d65831734b433c74aab52ab62e
// Content-Type: application/json
// 
// {"_id":"multipart","_rev":"4-5a99a8faccb6b0a8d6fd48cf1383ecee","_attachments":{"three.txt":{"content_type":"text/plain","revpos":4,"digest":"md5-E+kO+eF6BqbZs7CA3w0x4w==","length":52,"follows":true,"encoding":"gzip","encoded_length":69},"two.txt":{"content_type":"text/plain","revpos":3,"digest":"md5-MpEhbPHIFLQQJrX86NPA1Q==","length":39,"follows":true,"encoding":"gzip","encoded_length":59},"one.txt":{"content_type":"text/plain","revpos":2,"digest":"md5-mIz2XUNuIuC6D2VKfPc2YA==","length":30,"follows":true,"encoding":"gzip","encoded_length":50}}}
// --7c5ea2d65831734b433c74aab52ab62e
// Content-Disposition: attachment; filename="three.txt"
// Content-Type: text/plain
// Content-Length: 69
// Content-Encoding: gzip
// 
// <binary data>
// --7c5ea2d65831734b433c74aab52ab62e
// Content-Disposition: attachment; filename="two.txt"
// Content-Type: text/plain
// Content-Length: 59
// Content-Encoding: gzip
// 
// <binary data>
// --7c5ea2d65831734b433c74aab52ab62e
// Content-Disposition: attachment; filename="one.txt"
// Content-Type: text/plain
// Content-Length: 50
// Content-Encoding: gzip
// 
// <binary data>
// --7c5ea2d65831734b433c74aab52ab62e--
// ```
export default class MultipartUnpacker {
  constructor (contentType, onPart) {
    // parse boundary from content type header
    const [_, boundary] = contentType.match(/boundary="([^"]+)"/)
    const enc = new TextEncoder()
    this.boundary = enc.encode(`--${boundary}`)

    this.onPart = onPart

    this.decoder = new TextDecoder()

    this.data = new Uint8Array(0)
  }

  init () {
    return init()
  }

  read (chunk) {
    if (chunk) {
      // add current chunk to data
      const newData = new Uint8Array(this.data.length + chunk.length)
      newData.set(this.data, 0)
      newData.set(chunk, this.data.length)
      this.data = newData
    }
    
    // and parse what we already have
    this.parseParts()
  }

  // parse parts out of our data and store the rest
  parseParts () {
    while (true) {
      // get position of first boundary
      const startPosition = firstMatch(this.boundary, this.data)
      if (startPosition === -1) break
      
      // get position of the next boundary
      const endPosition = firstMatch(this.boundary, this.data, startPosition + this.boundary.length - 1)
      if (endPosition === -1) break
      
      // extract the data in between the boundaries
      const partData = this.data.slice(startPosition + this.boundary.length, endPosition)
      
      // parse that part
      this.parsePart(partData)
      
      // store the rest
      this.data = this.data.slice(endPosition)
    }
  }

  // parse one part. Can be either a doc eg:
  // ```
  // Content-Type: application/json
  // 
  // {"_id":"multipart","_rev":"4-5a99a8faccb6b0a8d6fd48cf1383ecee","_attachments":{"three.txt":{"content_type":"text/plain","revpos":4,"digest":"md5-E+kO+eF6BqbZs7CA3w0x4w==","length":52,"follows":true,"encoding":"gzip","encoded_length":69},"two.txt":{"content_type":"text/plain","revpos":3,"digest":"md5-MpEhbPHIFLQQJrX86NPA1Q==","length":39,"follows":true,"encoding":"gzip","encoded_length":59},"one.txt":{"content_type":"text/plain","revpos":2,"digest":"md5-mIz2XUNuIuC6D2VKfPc2YA==","length":30,"follows":true,"encoding":"gzip","encoded_length":50}}}
  // ```
  // or an attachment:
  // ```
  // Content-Disposition: attachment; filename="three.txt"
  // Content-Type: text/plain
  // Content-Length: 69
  // Content-Encoding: gzip
  // 
  // <binary data>
  // ```
  parsePart (data) {
    // get the position of the content after the header
    // header is separated by two blank lines
    const contentPosition = firstMatch(HEADER_SEPARATOR, data)
    if (contentPosition === -1) throw new Error('Wrong content position')

    // parse the header
    const headerData = data.slice(2, contentPosition)
    const header = this.decoder.decode(headerData)
    const headers = header.split('\r\n').reduce((memo, line) => {
      const [name, value] = line.split(/:\s*/)
      memo[name] = value
      return memo
    }, {})

    // content type is required
    const contentType = headers['Content-Type']
    if (!contentType) throw new Error('Missing Content-Type header')
    
    // extract the actual data
    let contentData = data.slice(contentPosition + HEADER_SEPARATOR.length, data.length - 2)

    const contentDisposition = headers['Content-Disposition']
    if (!contentDisposition && contentType === 'application/json') {
      // if there is no content disposition header and its a json, decode it
      const text = this.decoder.decode(contentData)
      const json = JSON.parse(text)
      this.onPart({
        json
      })
    } else {
      // if there is a content disposition, its an attachment
      // parse content disposition attributes
      const [type, attributesString] = contentDisposition.split(/;\s*/, 2)
      const attributes = attributesString.split(/;\s*/).reduce((memo, pair) => {
        const [name, value] = pair.split('=')
        memo[name] = JSON.parse(value)
        return memo
      }, {})

      const contentEncoding = headers['Content-Encoding']
      if (contentEncoding) {
        contentData = decompressGzip(contentData)
      }
      // construct a blob out of the binary data
      const blob = new Blob([contentData], { type: contentType })
      this.onPart({
        ...attributes,
        blob
      })
    }
  }
}
